storage/bootstrap/bootstrapper/commitlog/source.go

```diff
@@ -166,93 +249,187 @@ func (s *commitLogSource) Read(
-	emptyErrs := 0
-	bootstrapResult := result.NewBootstrapResult()
-	blocksPool := bopts.DatabaseBlockOptions().DatabaseBlockPool()
-	multiReaderIteratorPool := blopts.MultiReaderIteratorPool()
 	for shard, unmergedShard := range unmerged {
-		shardResult := result.NewShardResult(len(unmergedShard), s.opts.ResultOptions())
-		for _, unmergedBlocks := range unmergedShard {
-			blocks := block.NewDatabaseSeriesBlocks(len(unmergedBlocks.encoders))
-			for start, unmergedBlock := range unmergedBlocks.encoders {
-				block := blocksPool.Get()
-				if len(unmergedBlock) == 0 {
-					emptyErrs++
-					continue
-				} else if len(unmergedBlock) == 1 {
-					block.Reset(start, unmergedBlock[0].enc.Discard())
-				} else {
-					readers := make([]io.Reader, len(unmergedBlock))
-					for i := range unmergedBlock {
-						stream := unmergedBlock[i].enc.Stream()
-						readers[i] = stream
-						defer stream.Finalize()
-					}
+		mergeSemaphore <- true
+		wg.Add(1)
+		go func(shard int, unmergedShard encodersAndRanges) {
+			var shardResult result.ShardResult
+			shardResult, shardEmptyErrs[shard], shardErrs[shard] = s.mergeShard(
+				unmergedShard, blocksPool, multiReaderIteratorPool, encoderPool, blopts)
+			if len(shardResult.AllSeries()) > 0 {
+				// Prevent race conditions while updating bootstrapResult from multiple go-routines
+				bootstrapResultLock.Lock()
+				// Shard is a slice index so conversion to uint32 is safe
+				bootstrapResult.Add(uint32(shard), shardResult, nil)
+				bootstrapResultLock.Unlock()
+			}
+			<-mergeSemaphore
+			wg.Done()
+		}(shard, unmergedShard)
+	}
 
-					iter := multiReaderIteratorPool.Get()
-					iter.Reset(readers)
-
-					var err error
-					enc := encoderPool.Get()
-					enc.Reset(start, blopts.DatabaseBlockAllocSize())
-					for iter.Next() {
-						dp, unit, annotation := iter.Current()
-						encodeErr := enc.Encode(dp, unit, annotation)
-						if encodeErr != nil {
-							err = encodeErr
-							errs++
-							break
-						}
-					}
-					if iterErr := iter.Err(); iterErr != nil {
-						if err == nil {
-							err = iter.Err()
-						}
-						errs++
-					}
+	// Wait for all merge goroutines to complete
+	wg.Wait()
+	s.logMergeShardsOutcome(shardErrs, shardEmptyErrs)
+	return bootstrapResult
+}
 
-					iter.Close()
-					for i := range unmergedBlock {
-						unmergedBlock[i].enc.Close()
-					}
+func (s *commitLogSource) mergeShard(
+	unmergedShard encodersAndRanges,
+	blocksPool block.DatabaseBlockPool,
+	multiReaderIteratorPool encoding.MultiReaderIteratorPool,
+	encoderPool encoding.EncoderPool,
+	blopts block.Options,
+) (shardResult result.ShardResult, numShardEmptyErrs int, numErrs int) {
+	shardResult = result.NewShardResult(len(unmergedShard.encodersBySeries), s.opts.ResultOptions())
+	for _, unmergedBlocks := range unmergedShard.encodersBySeries {
+		blocks := block.NewDatabaseSeriesBlocks(len(unmergedBlocks.encoders))
+		for start, unmergedBlockEncoders := range unmergedBlocks.encoders {
```

Comment:
could you refactor this inner loop into smaller methods. E.g.

```
type []encoder encoders

func (e encoders) merge(...) (block.Block, error) {
// ...
}
```

Reply:
I managed to factor out the innermost loop into a new function called mergeSeries(). It's still a fairly long method, but I removed a lot of the identation and its a lot easier to read now I think

Reply:
Yeah it's way easier to grok